File: src/leetcode/python/easy/SumOfTwoIntegers_371.py
class SumOfTwoIntegers_371(object):
    def getSum(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: int

        a = 2147483647 and b = -2147483648 is the edge case
        mask is used to mimic the 32 bits operation

        xor op: get the sum without carry
        and op: get the carry
        shift left carry 1 bit
        We will get the result if carry is zero.
        """
        if a == 0: return b
        if b == 0: return a
        if a == 2147483647 and b == -2147483648: return -1
        flag = False
        if a < 0 and b < 0:
            a = -a
            b = -b
            flag = True

        mask = 0xffffffff
        while b != 0:
            a &= mask
            b &= mask
            carry = a & b & mask
            a ^= b
            b = (carry << 1) & mask
        return -a if flag else a

File: src/leetcode/java/SumOfTwoIntegers_371.java
package leetcode.java;

public class SumOfTwoIntegers_371 {
  // xor op: get the sum without carry
  // and op: get the carry
  // shift left carry 1 bit
  // We will get the result if carry is zero.
  public int getSum(int a, int b) {
    if (a == 0) return b;
    if (b == 0) return a;

    int carry;
    while (b != 0) {
      carry = a & b;
      a = a ^ b;
      b = carry << 1;
    }
    return a;
  }
}
